File: minecommons-core/src/main/java/com/gitlab/aecsocket/minecommons/core/effect/Effector.java
package com.gitlab.aecsocket.minecommons.core.effect;

import com.gitlab.aecsocket.minecommons.core.vector.cartesian.Vector3;

/**
 * An audience-like object which can accept an effect.
 */
public interface Effector {
    /**
     * Gets an effector that does nothing.
     * @return The effector.
     */
    static Effector empty() {
        return EmptyEffector.INSTANCE;
    }

    /**
     * Plays a sound effect.
     * @param effect The sound effect.
     * @param origin The origin position of the sound.
     */
    void play(SoundEffect effect, Vector3 origin);

    /**
     * Spawns a particle effect.
     * @param effect The particle effect.
     * @param origin The origin position of the particle.
     */
    void spawn(ParticleEffect effect, Vector3 origin);
}

File: core/src/main/kotlin/com/gitlab/aecsocket/alexandria/core/effect/Effector.kt
package com.gitlab.aecsocket.alexandria.core.effect

import com.gitlab.aecsocket.alexandria.core.physics.Vector3

interface Effector {
    fun playSound(effect: SoundEffect, position: Vector3)

    fun showParticle(effect: ParticleEffect, position: Vector3)

    companion object {
        val EMPTY: Effector = EmptyEffector
    }
}

private object EmptyEffector : Effector {
    override fun playSound(effect: SoundEffect, position: Vector3) {}
    override fun showParticle(effect: ParticleEffect, position: Vector3) {}
}

fun interface ForwardingEffector : Effector {
    fun effectors(): Iterable<Effector>

    override fun playSound(effect: SoundEffect, position: Vector3) =
        effectors().forEach { it.playSound(effect, position) }

    override fun showParticle(effect: ParticleEffect, position: Vector3) =
        effectors().forEach { it.showParticle(effect, position) }
}

fun Iterable<SoundEffect>.play(effector: Effector, position: Vector3) =
    forEach { effector.playSound(it, position) }

fun Iterable<ParticleEffect>.show(effector: Effector, position: Vector3) =
    forEach { effector.showParticle(it, position) }
